src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SSRP.cs

```diff
@@ -0,0 +1,151 @@
+        /// <returns>DAC port for given instance name</returns>
+        internal static int GetDacPortByInstanceName(string browserHostName, string instanceName)
+        {
+            Debug.Assert(!string.IsNullOrWhiteSpace(browserHostName), "browserHostName should not be null, empty, or whitespace");
+            Debug.Assert(!string.IsNullOrWhiteSpace(instanceName), "instanceName should not be null, empty, or whitespace");
+
+            byte[] dacPortInfoRequest = CreateDacPortInfoRequest(instanceName);
+            byte[] responsePacket = SendUDPRequest(browserHostName, SqlServerBrowserPort, dacPortInfoRequest);
+
+            const byte SvrResp = 0x05;
+            const byte ProtocolVersion = 0x01;
+            const byte RespSize = 0x06;
+            if (responsePacket == null || responsePacket.Length <= 4 || responsePacket[0] != SvrResp ||
+                BitConverter.ToUInt16(responsePacket, 1) != RespSize || responsePacket[3] != ProtocolVersion)
+            {
+                throw new SocketException();
+            }
+
+            int dacPort = BitConverter.ToUInt16(responsePacket, 4);
+            return dacPort;
+        }
+
+        /// <summary>
+        /// Creates DAC port lookup request (CLNT_UCAST_DAC) for given instance name.
+        /// </summary>
+        /// <param name="instanceName">instance name to lookup DAC port</param>
+        /// <returns>Byte array of DAC port lookup request (CLNT_UCAST_DAC)</returns>
+        private static byte[] CreateDacPortInfoRequest(string instanceName)
+        {
+            Debug.Assert(!string.IsNullOrWhiteSpace(instanceName), "instanceName should not be null, empty, or whitespace");
+
+            const byte ClntUcastDac = 0x0F;
+            const byte ProtocolVersion = 0x01;
+            int byteCount = Encoding.ASCII.GetByteCount(instanceName);
+
+            byte[] requestPacket = new byte[byteCount + 2];
+            requestPacket[0] = ClntUcastDac;
+            requestPacket[1] = ProtocolVersion;
+            Encoding.ASCII.GetBytes(instanceName, 0, instanceName.Length, requestPacket, 2);
+
+            return requestPacket;
+        }
+
+        /// <summary>
+        /// Sends request to server, and receives response from server by UDP.
+        /// </summary>
+        /// <param name="browserHostname">UDP server hostname</param>
+        /// <param name="port">UDP server port</param>
+        /// <param name="requestPacket">request packet</param>
+        /// <returns>response packet from UDP server</returns>
+        private static byte[] SendUDPRequest(string browserHostname, int port, byte[] requestPacket)
+        {
+            Debug.Assert(!string.IsNullOrWhiteSpace(browserHostname), "browserhostname should not be null, empty, or whitespace");
+            Debug.Assert(port >= 0 && port <= 65535, "Invalide port range");
+            Debug.Assert(requestPacket != null && requestPacket.Length > 0, "requestPacket should not be null or 0-length array");
```

Comment:
The `Assert(Boolean, String)` definition is 
```
Checks for a condition; if the condition is *false*, outputs a specified message and displays a message box that shows the call stack.
```
The message send to listner when the condition returns *false*.
https://msdn.microsoft.com/en-us/library/system.diagnostics.debug.assert(v=vs.110).aspx